### PR TITLE
use a bulk command to delete unneeded files, instead of processing one at time

### PR DESCRIPTION
**Pull request by woodhull**
in our environment, one at a time deletion is much slower, causes the ruby process to use large amounts of memory. This call is fast and efficient instead. 

Batches in groups of 500 to avoid issues with enormous requests.

### CODE REVIEW

File: lib/asset_sync/storage.rb [hunk 1]
```diff
@@ -106,13 +106,12 @@ def delete_file(f, remote_files_to_delete)
     def delete_extra_remote_files
       log "Fetching files to flag for delete"
       remote_files = get_remote_files
-      # fixes: https://github.com/rumblelabs/asset_sync/issues/19
```
Comment (PikachuEXE) [inline lib/asset_sync/storage.rb]: Please revert the removal of comments
It's unrelated to this PR

File: spec/unit/storage_spec.rb [hunk 2]
```diff
@@ -307,4 +307,25 @@ def check_file(file)
...
+      remote_files = ['public/image.png']
+
+      storage = AssetSync::Storage.new(@config)
+
+      [:local_files, :ignored_files, :always_upload_files].each do |method|
+         expect(storage).to receive(method).and_return([])
+      end
+
+      expect(storage).to receive(:get_remote_files).and_return(remote_files)
+      expect(storage).to receive(:connection).and_return(connection = double)
```
Comment (PikachuEXE) [inline spec/unit/storage_spec.rb]: Maybe can assign these variable above?
Inline assignments are easier to be overlook

File: lib/asset_sync/storage.rb [hunk 3]
```diff
@@ -106,13 +106,12 @@ def delete_file(f, remote_files_to_delete)
...
+      remote_files_to_delete = remote_files - local_files - ignored_files - always_upload_files
 
-      log "Flagging #{from_remote_files_to_delete.size} file(s) for deletion"
-      # Delete unneeded remote files
-      bucket.files.each do |f|
-        delete_file(f, from_remote_files_to_delete)
+      log "Flagging #{remote_files_to_delete.size} file(s) for deletion"
+      # Delete unneeded remote files, in chunks of 500
+      remote_files_to_delete.each_slice(500) do |slice|
+        connection.delete_multiple_objects(self.config.fog_directory, slice)
```
Comment (PikachuEXE) [inline lib/asset_sync/storage.rb]: Is there any doc explaining whether `delete_multiple_objects` is a public API and it's usage?
Also it seems not every storage have it implemented

**Comment by JanStevens** [comment]
Anyway we can get this in asset sync? Or pipelines take considerable longer because asset sync deletes files one by one. I can take up the MR but only if it stands a chance in getting merged.

We are using S3 storage

**Comment by PikachuEXE** [comment]
@JanStevens 
I think my comments still hold
I have no idea what `delete_multiple_objects` does or is it available for all storage providers this gem supports
You can help by providing more details

**Comment by JanStevens** [comment]
I could adapt the code to see if the provider responds to `delete_multiple_objects` ? If not we use the old code one by one, if it does then we use `delete_multiple_objects`. There is probably a fog way to figure out if an action is supported

I see that S3 support `delete_multiple_objects` but azure not.

**Comment by PikachuEXE** [comment]
If that's the only way to detect support of `delete_multiple_objects`, I am fine with it
But another concern would be whether all providers got same interface for `delete_multiple_objects` 
(whether `delete_multiple_objects` from different providers expect same arguments)

Also I think it's easier to "whitelist" the provider to use `delete_multiple_objects`
```ruby
# Might not be exact code
if self.config.aws? && connection.respond_to?(:delete_multiple_objects)
  # Use `delete_multiple_objects`
else
  # Current code
end
```

If you can come up with a PR like this and you tested it yourself I think I can merge it soon
After next week I am on vacation for 2 weeks

**Comment by PikachuEXE** [comment]
Closing this since #392 merged

**Pull request closed without merging**